File: src/main/java/ir/reservs/reservs/ui/main/history/HistoryPresenter.java
package ir.reservs.reservs.ui.main.history;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import ir.reservs.reservs.data.DataManager;
import ir.reservs.reservs.utils.RetrofitError;

public class HistoryPresenter implements HistoryContract.Presenter {

    private DataManager dataManager;
    private CompositeDisposable compositeDisposable;
    private HistoryContract.View view;

    @Inject
    public HistoryPresenter(DataManager dataManager, CompositeDisposable compositeDisposable) {
        this.dataManager = dataManager;
        this.compositeDisposable = compositeDisposable;
    }

    public void getDataHistory() {
        view.loadingState();
        Disposable disposable = dataManager.reserves()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        reserves -> {
                            view.normalState();
                            view.setHistoryData(reserves);
                        },
                        error -> {
                            RetrofitError.INSTANCE.handle(view, error);
                            view.errorState();
                        }
                );
        compositeDisposable.add(disposable);
    }

    @Override
    public void onAttach(HistoryContract.View view) {
        this.view = view;

    }

    @Override
    public void onDetach() {
        view = null;
        if (!compositeDisposable.isDisposed()) {
            compositeDisposable.clear();
        }
    }
}

File: src/main/java/ir/reservs/reservs/ui/main/history/HistoryPresenter.kt
package ir.reservs.reservs.ui.main.history

import io.reactivex.disposables.CompositeDisposable
import ir.reservs.reservs.data.DataManager
import ir.reservs.reservs.utils.RetrofitError

class HistoryPresenter(val dataManager: DataManager,
                       private val compositeDisposable: CompositeDisposable)
    : HistoryContract.Presenter {
    private var page = 1
    private var isLastPage = false
    private var isLoading = false
    private var view: HistoryContract.View? = null

    override fun onAttach(view: HistoryContract.View) {
        this.view = view
        fetchHistories()
    }

    fun fetchHistories() {
        isLoading = true
        val d = dataManager.reserves(page)
                .subscribe({
                    isLoading = false
                    page++
                    view?.updateAdapter(it)
                    if (it.size > 0)
                        view?.normalState()
                    else {
                        if (page == 1)
                            view?.emptyState()
                        else
                            isLastPage = true
                    }
                }, {
                    isLoading = false
                    RetrofitError.handle(view!!, it)
                    view?.errorState()
                })
        compositeDisposable.add(d)
    }

    fun isLastPage(): Boolean {
        return isLastPage
    }

    fun isLoading(): Boolean {
        return isLoading
    }

    override fun onDetach() {
        view = null
        if (!compositeDisposable.isDisposed) {
            compositeDisposable.clear()
        }
    }
}
